### DISCUSSION TITLE
Has anyone measured difference between optimized Tailwindcss and plain CSS gzipped sizes for the same page design?

### DISCUSSION THREAD

**Discussion by kawpii** [tailwindlabs/tailwindcss#5222] [Help] [answered]
I know that Tailwindcss size could be reduce by purging CSS, but has anyone measured the gzipped sizes of optimized Tailwindcss and plain CSS for same page design? Definitely there is a overhead with Tailwindcss, but how much is it (as a percentage)?

**Comment by olets** [comment]
> Definitely there is a overhead with Tailwindcss

Any utility class solution is almost certain to result in a _smaller_ final stylesheet. Consider this very minimal example:

```html
<div style="display: block">component 1 content</div>
<div style="display: block">component 2 content</div>
```

If not using utility classes you'd likely end up with two selectors, something like

```html
<div class="component-1">component 1 content</div>
<div class="component-2">component 2 content</div>
```

```css
/* after minification */
.component-1,.component-2{display:block}
```

If using a utility class you'd end up with something like

```html
<div class="block">component 1 content</div>
<div class="block">component 2 content</div>
```

```css
/* after minification */
.block{display:block}
```

> how much is it (as a percentage)

This depends entirely on the design. The more reuse of CSS rules, the bigger the savings should be when you use utility classes. If you use Tailwind CSS and use a lot of `@apply` you'll miss out on some of that weight reduction. And plugins also have the potential to make the final stylehsheet heavier.

**Reply by kawpii** [reply]
Instead of 
```
.component-1,.component-2{display:block}

<div class="component-1">component 1 content</div>
<div class="component-2">component 2 content</div>
```

We can rewrite this a component

```
.component{display:block}

<div class="component">component 1 content</div>
<div class="component">component 2 content</div>
```

Say if you want change the background color you do

```
/* 49 characters */
.component{display:block;background-color:green;}
```

with utility approach you do something like


```
<div class="block bg-green">component 1 content</div>
<div class="block bg-green">component 2 content</div>

/* 56 characters */
.block{display:block}.bg-green{background-color:green;}
```

So there is tiny penalty with utility approach, but this is going to negligible with compression as @simonswiss pointed out.

Another thing to note here is those differences are caused by class names (1 class name vs 2 class names) and additional curly braces from utility classes.

**Reply by olets** [reply]
Yep! There are additional factors that will make the final result entirely dependent on your design and CSS style.

The assumption in my naming those "components" is that they wouldn't be refactored across, and that there'd be more to each of them than the single rule. Imagine two components with separate stylesheets, one is display block and four other rules, another is display block and ten other rules.

You're right about class name length too. The impact of class will be different depending on how you name things. My team traditionally used BEM, where class names can get real long real fast. Even with short names and extra braces, the CSS weight benefit of utility classes should increase as the amount of duplication it eliminates. Imagine a design where all the site's styles appear above the fold, and the rest of the page reuses those styles but in different combos (doesn't havehave to be too special design, just one with a reliable design system)... That whole rest of the page might not contribute to CSS weight at all!

If you end up doing the experiment on one of your projects it'd be interesting if to shared your findings here.

**Comment by simonswiss** [comment, marked as answer]
Check this fascinating article by Nicolas Gallagher (from 2012, but still **extremely** relevant!):

http://nicolasgallagher.com/about-html-semantics-front-end-architecture/

Near the end, the **A note on raw file size and HTTP compression** section makes this exact experiment, to see the "weight" of utility-class based approaches on HTML files.

The results will likely please you 👍

I did the same experiment on some projects of mine, and observed the same thing.

![image](https://user-images.githubusercontent.com/485747/129654347-ea0b882a-3a38-4c1a-9eac-c4bd9c457a13.png)
